compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -1191,67 +1158,314 @@ trait Implicits { self: Typer =>
+ * recursive references and emit a complete implicit dictionary when the outermost search
+ * is complete.
  */
-class SearchHistory(val searchDepth: Int, val seen: Map[ClassSymbol, Int]) {
-
-  /** The number of applications and refinements in this type, after all aliases are expanded */
-  private def typeSize(tp: Type)(implicit ctx: Context): Int = {
-    val accu = new TypeAccumulator[Int] {
-      def apply(n: Int, tp: Type): Int = tp match {
-        case tp: AppliedType =>
-          foldOver(n + 1, tp)
-        case tp: RefinedType =>
-          foldOver(n + 1, tp)
-        case tp: TypeRef if tp.info.isTypeAlias =>
-          apply(n, tp.superType)
-        case _ =>
-          foldOver(n, tp)
+abstract class SearchHistory { outer =>
+  val root: SearchRoot
+  val open: List[(Candidate, Type)]
+  /** Does this search history contain any by name implicit arguments. */
+  val byname: Boolean
+
+  /**
+   * Create the state for a nested implicit search.
+   * @param cand The candidate implicit to be explored.
+   * @param pt   The target type for the above candidate.
+   * @result     The nested history.
+   */
+  def nest(cand: Candidate, pt: Type)(implicit ctx: Context): SearchHistory = {
+    new SearchHistory {
+      val root = outer.root
+      val open = (cand, pt) :: outer.open
+      val byname = outer.byname || isByname(pt)
+    }
+  }
+
+  def isByname(tp: Type): Boolean = tp.isInstanceOf[ExprType]
+
+  /**
+   * Check if the supplied candidate implicit and target type indicate a diverging
+   * implicit search.
+   *
+   * @param cand The candidate implicit to be explored.
+   * @param pt   The target type for the above candidate.
+   * @result     True if this candidate/pt are divergent, false otherwise.
+   */
+  def checkDivergence(cand: Candidate, pt: Type)(implicit ctx: Context): Boolean = {
+    // For full details of the algorithm see the SIP:
+    //   https://docs.scala-lang.org/sips/byname-implicits.html
+
+    val widePt = pt.widenExpr
+    lazy val ptCoveringSet = widePt.coveringSet
+    lazy val ptSize = widePt.typeSize
+    lazy val wildPt = wildApprox(widePt)
+
+    // Unless we are able to tie a recursive knot, we report divergence if there is an
+    // open implicit using the same candidate implicit definition which has a type which
+    // is larger (see `typeSize`) and is constructed using the same set of types and type
+    // constructors (see `coveringSet`).
+    //
+    // We are able to tie a recursive knot if there is compatible term already under
+    // construction which is separated from this context by at least one by name argument
+    // as we ascend the chain of open implicits to the outermost search context.
+
+    @tailrec
+    def loop(ois: List[(Candidate, Type)], belowByname: Boolean): Boolean = {
+      ois match {
+        case Nil => false
+        case (hd@(cand1, tp)) :: tl =>
+          (if (cand1.ref == cand.ref) {
```

Comment:
Why not fuse the scrutinee and the match? I.e.
```scala
if (cand1.ref == cand.ref) {
  val wideTp = tp.widenExpr
  lazy val wildTp = wildApprox(wideTp)
  if (belowByname && (wildTp <:< wildPt)) false
  else if ((wideTp.typeSize < ptSize && wideTp.coveringSet == ptCoveringSet) || (wildTp == wildPt)) true
  else loop(tl, isByname(tp) || belowByname)
}
else loop(tl, isByname(tp) || belowByname)
```
The recursive call is duplicated, but everything else is shorter and nicer (and less boxing, too).

Reply:
Agreed.